core/src/main/java/hudson/PluginManager.java

```diff
@@ -1341,41 +1343,55 @@ public static boolean isNonMetaLabel(String label) {
         List<JSONObject> plugins = new ArrayList<>();
         for (UpdateSite site : Jenkins.get().getUpdateCenter().getSiteList()) {
             List<JSONObject> sitePlugins = site.getAvailables().stream()
                 .filter(plugin -> {
-                    if (StringUtils.isBlank(query)) {
+                    // preserve an existing feature for the "warning:" query
+                    if (isWarningsQuery && plugin.hasWarnings()) {
                         return true;
                     }
-                    return StringUtils.containsIgnoreCase(plugin.name, query) ||
-                        StringUtils.containsIgnoreCase(plugin.title, query) ||
-                        StringUtils.containsIgnoreCase(plugin.excerpt, query) ||
-                        plugin.hasCategory(query) ||
-                        plugin.getCategoriesStream()
-                            .map(UpdateCenter::getCategoryDisplayName)
-                            .anyMatch(category -> StringUtils.containsIgnoreCase(category, query)) ||
-                        plugin.hasWarnings() && query.equalsIgnoreCase("warning:");
-                })
-                .limit(Math.max(limit - plugins.size(), 1))
-                .sorted((o1, o2) -> {
-                    String o1DisplayName = o1.getDisplayName();
-                    if (o1.name.equalsIgnoreCase(query) ||
-                        o1DisplayName.equalsIgnoreCase(query)) {
-                        return -1;
-                    }
-                    String o2DisplayName = o2.getDisplayName();
-                    if (o2.name.equalsIgnoreCase(query) || o2DisplayName.equalsIgnoreCase(query)) {
-                        return 1;
+                    // if the query is a category, return plugins in that category
+                    if (plugin.hasCategory(_query)) {
+                        return true;
                     }
-                    if (o1.name.equals(o2.name)) {
-                        return 0;
+                    // otherwise we need to find a specific plugin based on the query
+                    // for short queries, we want plugins whose title or name starts with the query
+                    // for longer, multi-word queries, we want any part of the title to match the query words
+                    final String _title = plugin.title.toLowerCase(Locale.ROOT);
+                    final String _name = plugin.name.toLowerCase(Locale.ROOT);
+                    switch(_query.length()) {
+                        case 0:
+                            return true;
+                        case 1:
+                        case 2:
+                        case 3: // maybe go up to 4 or 5?
+                            return _title.startsWith(_query)
+                                    || _name.startsWith(_query);
+                        default:
+                            List<String> _firstQueryWords = new LinkedList<>(Arrays.asList(_query.split("\\s+")));
```

Comment:
what's with all these `_` at the start of variable names?

Reply:
`_title` means "lowercase title." I rarely use `_xyz` variable names, but the method signature has a `query`, which I converted to `_query` ("lowercase query"), and that started a trend..